src/fabric/src/fabric_util.erl

```diff
@@ -136,24 +136,36 @@ get_shard([#shard{node = Node, name = Name} | Rest], Opts, Timeout, Factor) ->
     MFA = {fabric_rpc, open_shard, [Name, [{timeout, Timeout} | Opts]]},
     Ref = rexi:cast(Node, self(), MFA, [sync]),
     try
-        receive
-            {Ref, {ok, Db}} ->
-                {ok, Db};
-            {Ref, {'rexi_EXIT', {{unauthorized, _} = Error, _}}} ->
-                throw(Error);
-            {Ref, {'rexi_EXIT', {{forbidden, _} = Error, _}}} ->
-                throw(Error);
-            {Ref, Reason} ->
-                couch_log:debug("Failed to open shard ~p because: ~p", [Name, Reason]),
-                get_shard(Rest, Opts, Timeout, Factor)
-        after Timeout ->
-            couch_log:debug("Failed to open shard ~p after: ~p", [Name, Timeout]),
-            get_shard(Rest, Opts, Factor * Timeout, Factor)
-        end
+        await_shard_response(Ref, Name, Rest, Opts, Factor, Timeout)
     after
         rexi_monitor:stop(Mon)
     end.
 
+await_shard_response(Ref, Name, Rest, Opts, Factor, Timeout) ->
+    receive
+        Msg0 ->
+            {Msg, Delta} = rexi_utils:extract_delta(Msg0),
+            couch_stats_resource_tracker:accumulate_delta(Delta),
+            case Msg of
+                {Ref, {ok, Db}} ->
+                    {ok, Db};
+                {Ref, {'rexi_EXIT', {{unauthorized, _} = Error, _}}} ->
+                    throw(Error);
+                {Ref, {'rexi_EXIT', {{forbidden, _} = Error, _}}} ->
+                    throw(Error);
+                {Ref, Reason} ->
+                    couch_log:debug("Failed to open shard ~p because: ~p", [Name, Reason]),
+                    get_shard(Rest, Opts, Timeout, Factor);
+                %% {OldRef, {ok, Db}} -> <-- stale db resp that got here late, should we do something?
+                _ ->
+                    %% Got a message from an old Ref that timed out, try again
+                    await_shard_response(Ref, Name, Rest, Opts, Factor, Timeout)
```

Comment:
This changes the logic a bit to drop any messages. This doesn't seem to be related to the stats tracking, but it's a good change and might be better do it as a separate PR to discuss there.

This goes back to the effort some years back to use {rexi, ...} prefix for rexi messages to avoid dropping or consuming any 2 items tuple message. Wonder if we should finish that work first and incorporate message cleanup with it. But overall it does seems like a problem for a different PR.